main.py

```diff
@@ -1,405 +1,519 @@
+                        display_time = 10
+                        AI_MODE = not AI_MODE
+                        ai_reset()
+                        valid_moves, turn = [], None
+                        #square_selected = ()
+                       # player_clicks = []
+                        valid_moves, first_click_turn = gs.get_valid_moves()
+                        print("AI MODE ENABLED") if AI_MODE else print(
+                            "AI MODE DISABLED")
+
+                    elif e.key == pg.K_p and not AI_MODE and game_over:
+                        PLAYBACK_MODE = not PLAYBACK_MODE
+                        playback_log = gs.move_log
+                        gs = Game_state()
+                        valid_moves, turn = [], None
+                        square_selected = ()
+                        player_clicks = []
+                        valid_moves, first_click_turn = gs.get_valid_moves()
+                        print("PLAYBACK MODE ENABLED \nNumber of Moves Available: " +
+                              str(len(playback_log))) if PLAYBACK_MODE else print(
+                            "PLAYBACK MODE DISABLED")
+                        playback_index = 0
+                    elif e.key == pg.K_n and PLAYBACK_MODE and game_over:
+                        if len(playback_log) == 0:
+                            print("No Moves to Play")
+                            break
+                        if playback_index == len(playback_log):
+                            #playback_index = len(playback_log) - 1
+                            print("Max Playback Reached!")
+                        else:
+                            str_sqr = str(playback_log[playback_index])
+                            start_row = int(str_sqr[1])
+                            start_col = int(str_sqr[4])
+                            start_sqr = (start_row, start_col)
+                            end_row = int(str_sqr[8])
+                            end_col = int(str_sqr[11])
+                            end_sqr = (end_row, end_col)
+                            clicked_sqr = [start_sqr, end_sqr]
+                            # print(start_sqr)
+                            # print(end_sqr)
+                            # print(str_sqr)
+                            # print(clicked_sqr)
+                            playback_index += 1
+                            move = Move(
+                                start_sqr, end_sqr, gs.board)
+                            gs.make_move(move)
+                            animate(move, screen, gs.board, clock)
+                            print("Move: "+str((playback_index)))
+                            print(move.get_chess_notation())
+
+                    elif e.key == pg.K_b and PLAYBACK_MODE and game_over:
+                        if playback_index <= 0:
+                            playback_index = 0
+                            print("Min Playback Reached!")
+
+                        else:
```

Comment:
your error arises because when there is a capture in playback.. it does'nt undo the capture... i suggest you should use the undo function to make the previous move during playback

Reply:
Fixed the "KeyError" Issue